fullnode/fullnode_cache.go

```diff
@@ -0,0 +1,228 @@
+}
+
+// Prefix for cache keys to avoid collisions with other keys.
+const (
+	sessionCacheKeyPrefix       = "session"
+	accountPubKeyCacheKeyPrefix = "pubkey"
+)
+
+// cachingFullNode wraps a LazyFullNode with SturdyC-based caching.
+// - Early refresh: background updates before expiry (prevents thundering herd/latency spikes)
+// - Example: 30s TTL, refresh at 22.5–27s (75–90%)
+// - Benefits: zero-latency reads, graceful degradation, auto load balancing
+// Docs: https://github.com/viccon/sturdyc
+type cachingFullNode struct {
+	logger polylog.Logger
+
+	// Underlying node for protocol data fetches
+	lazyFullNode *LazyFullNode
+
+	// Session cache (5 min on Beta TestNet, see #275)
+	// TODO_MAINNET_MIGRATION(@Olshansk): Revisit after mainnet
+	sessionCache *sturdyc.Client[sessiontypes.Session]
+
+	// The account public key cache; used to cache account public keys indefinitely.
+	// It has an infinite TTL and is populated only once on startup.
+	accountPubKeyCache *sturdyc.Client[cryptotypes.PubKey]
+}
+
+// NewCachingFullNode wraps a LazyFullNode with:
+//   - Session cache: refreshes early to avoid thundering herd/latency spikes
+//   - Account public key cache: indefinite cache for account data
+func NewCachingFullNode(
+	logger polylog.Logger,
+	lazyFullNode *LazyFullNode,
+	cacheConfig CacheConfig,
+) (*cachingFullNode, error) {
+	// Set default session TTL if not set
+	cacheConfig.hydrateDefaults()
+
+	// Log cache configuration
+	logger.Debug().
+		Str("cache_config_session_ttl", cacheConfig.SessionTTL.String()).
+		Msgf("cachingFullNode - Cache Configuration")
+
+	// Configure session cache with early refreshes
+	sessionMinRefreshDelay, sessionMaxRefreshDelay := getCacheDelays(cacheConfig.SessionTTL)
+
+	// Create the session cache with early refreshes
+	sessionCache := sturdyc.New[sessiontypes.Session](
+		cacheCapacity,
+		numShards,
+		cacheConfig.SessionTTL,
+		evictionPercentage,
+		// See: https://github.com/viccon/sturdyc?tab=readme-ov-file#early-refreshes
+		sturdyc.WithEarlyRefreshes(
+			sessionMinRefreshDelay,
+			sessionMaxRefreshDelay,
+			cacheConfig.SessionTTL,
+			retryBaseDelay,
+		),
+	)
+
+	// Create the account cache, which is effectively infinite
+	// caching for the lifetime of the application.
+	// No need to configure early refreshes with SturdyC.
+	accountPubKeyCache := sturdyc.New[cryptotypes.PubKey](
+		cacheCapacity,
+		numShards,
+		pubKeyCacheTTL,
+		evictionPercentage,
+	)
+
+	// Initialize the caching full node with the modified lazy full node
+	return &cachingFullNode{
+		logger:             logger,
+		lazyFullNode:       lazyFullNode,
+		sessionCache:       sessionCache,
+		accountPubKeyCache: accountPubKeyCache,
+	}, nil
+}
+
+// GetApp is a NoOp (apps fetched only at startup; relaying fetches sessions for app/session sync).
+func (cfn *cachingFullNode) GetApp(ctx context.Context, appAddr string) (*apptypes.Application, error) {
+	return nil, fmt.Errorf("GetApp is a NoOp in the caching full node")
```

Comment:
I disagree with this.

We should check if its in the cache.

If it's a miss -> we get it and store it.